source/blog/_posts/2016-10-12-promises.md

```diff
@@ -0,0 +1,353 @@
+---
+title: Promises - Asynchronous processes made easy
+tags:
+- javascript
+- promises
+- theme
+
+categories:
+- dev
+
+authors: [stp]
+
+github_link: blog/_posts/2016-10-12-promises.md
+---
+
+## Promises - Asynchronous processes made easy
+
+Javascript developers are knowing the struggle - you're having a bunch of asynchronous processes you have to time and use their results to call other methods. If it comes to XML HTTP requests or the FileReader API - asynchronous methods are everwhere in the JavaScript world. Without a sophisticated architecture you're probably ending up in the well-known "pyramid of doom" or "callback hell". You're nesting callbacks over callbacks which has a huge impact of the maintainability and adds a lot of dependencies to your code.
+
+```
+step1(function (value1) {
+   step2(value1, function(value2) {
+       step3(value2, function(value3) {
+           step4(value3, function(value4) {
+               // Do something with value4
+           });
+       });
+   });
+});
+```
+*Example for a typical "pyramid of doom"*
+
+`Promises` are a way around this problem. This blog post will take you on a ride to explore how promises are working and how you can use them right away. The goal is to get rid of the "pyramid of doom" and flatten the pyramid like this:
+
+```
+promise.then(step1)
+.then(step2)
+.then(step3)
+.then(step3)
+.then(function(value4) {
+    // Do something with value4
+})
+.catch(function(err) {
+   // Handle any error from all above steps
+});
+```
+*Example for a promise based approach*
+
+## Getting started with promises
+The technology is part of the ECMAScript 2015 (ES6) standard. The `Promise` object is used for asynchronous operations and has four states which are representing the current status of the operation:
+
+* `pending` - initial status, the promise is not `fulfilled` or `rejected` at this point
+* `fulfilled` - the operation was successful
+* `rejected` - the operation wasn't successful and an error occured
+* `settled` - the operation is not `fulfilled` or `rejected` and not `pending` anymore
+
+The `Promise` interface represents a proxy for a value which is not defined / known at the time the promise was defined. This allows an association of a `handler` method and the successful or failure of an asynchronous operation.
```

Comment:
Maybe only "success" in this case?
"[...] and the success or failure of an [...]"